tea.py

```diff
@@ -0,0 +1,460 @@
+        self.results["Negative electrode with electrolyte density [kg.m-3]"] = self.results.get("Negative electrode with electrolyte density [kg.m-3]") or self.results.get("Negative electrode composite weight fraction") * self.results.get("Negative electrode composite true density [kg.m-3]") + self.results.get("Electrolyte density [kg.m-3]") * (1 - self.results.get("Negative electrode composite weight fraction"))
+        self.results["Positive electrode with electrolyte density [kg.m-3]"] = self.results.get("Positive electrode with electrolyte density [kg.m-3]") or self.results.get("Positive electrode composite weight fraction") * self.results.get("Positive electrode composite true density [kg.m-3]") + self.results.get("Electrolyte density [kg.m-3]") * (1 - self.results.get("Positive electrode composite weight fraction"))
+        self.results["Separator with electrolyte density [kg.m-3]"] = self.results.get("Separator with electrolyte density [kg.m-3]") or self.results.get("Separator weight fraction") * self.results.get("Separator true density [kg.m-3]") + self.results.get("Electrolyte density [kg.m-3]") * (1 - self.results.get("Separator weight fraction"))
+
+        # stack density
+        self.results["Stack density [kg.m-3]"] = self.results.get("Stack density [kg.m-3]") or (self.results.get("Negative current collector thickness [m]") / 2 * self.results.get("Negative current collector density [kg.m-3]") + self.results.get("Negative electrode thickness [m]") * self.results.get("Negative electrode with electrolyte density [kg.m-3]") + self.results.get("Separator thickness [m]") * self.results.get("Separator with electrolyte density [kg.m-3]") + self.results.get("Positive electrode thickness [m]") * self.results.get("Positive electrode with electrolyte density [kg.m-3]") + self.results.get("Positive current collector thickness [m]") / 2 * self.results.get("Positive current collector density [kg.m-3]")) / self.results.get("Stack thickness [m]")
+        
+        # gravimetric stack energy density Wh.kg-1 better "specific energy"?
+        self.results["Gravimetric stack energy density [Wh.kg-1]"] = self.results.get("Gravimetric stack energy density [Wh.kg-1]") or self.results.get("Volumetric stack energy density [Wh.L-1]") / self.results.get("Stack density [kg.m-3]") * 1000
+        
+        if print_values == True:
+            print(self.results)
+
+        return self.results.get("Volumetric stack energy density [Wh.L-1]"), self.results.get("Gravimetric stack energy density [Wh.kg-1]")
+
+    def stack_breakdown(self, inputs=None, print_values=False):
+        # Inputs: "Electrolyte density [kg.m-3]", "Negative electrode active material true density [kg.m-3]", "Positive electrode active material true density [kg.m-3]"
+        if inputs is None:
+            inputs = {}
+        else:
+            inputs = inputs
+
+        # call stack_energy_densities
+        self.stack_energy_densities(inputs=inputs)
+
+        # volume loadings
```

Comment:
the or logic everywhere makes this hard to read and i think overcomplicated things?